### PR TITLE
Adding default list of commands to autocomplete

### PR DESCRIPTION
**Pull request by natestemen**
Clearing up issue listed here: https://github.com/overleaf/sharelatex/issues/51

### CODE REVIEW

File: public/coffee/ide/editor/directives/aceEditor/auto-complete/StaticSuggestionManager.coffee [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            meta: "cmd"
+        }
+    tripleArgumentCommands = for com in tripleArgumentCommands
+        {
+            caption: "\\#{com}{}{}{}"
+            snippet: "\\#{com}{$1}{$2}{$3}"
+            meta: "cmd"
+        }
+    special = for com in special
+        if com == 'def'
```
Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/StaticSuggestionManager.coffee]: `\def` and `\let` are both TeX commands, so I don't think we should be auto-completing them for LaTeX. `\newcommand` is the 'correct' way to do it (and we might even start actually linting at some point, and actively warning against things like `\def`).

**Comment by jpallen** [comment]
Generally looks good, thanks! 👍  Just a few comments

File: public/coffee/ide/editor/directives/aceEditor/auto-complete/AutoCompleteManager.coffee [hunk 2]
```diff
@@ -112,7 +112,10 @@ define [
 							else
 								callback null, result
 
-			@editor.completers = [@suggestionManager, SnippetCompleter, ReferencesCompleter, LabelsCompleter]
+			@editor.completers = [@suggestionManager,
```
Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/AutoCompleteManager.coffee]: Please can you indent this like:

```
foo = [
    one,
    two,
    ...
]
```

Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/AutoCompleteManager.coffee]: Please can you indent this like:

```
foo = [
    one,
    two,
    ...
]
```

File: public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee [hunk 3]
```diff
@@ -1,12 +1,83 @@
...
+                meta: "cmd"
+            }
+        else if com == 'LaTeX'
+            {
+                caption: "\\LaTeX{}"
+                snippet: "\\LaTeX{}"
+                meta: "cmd"
+            }
+
+    staticCommands = [].concat(noArgumentCommands,
```
Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: Similar for indentation

File: public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee [hunk 4]
```diff
@@ -95,7 +166,17 @@ define () ->
...
-	}]
+	}, {
+		caption: "\\begin{thebibliography}..."
+		snippet: """
+			\\begin{thebibliography}{$1}
+			\\bibitem{$2}
+			$3
+			\\end{thebibliography}
+		"""
+		}]
```
Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: Needs shifting back one indent.

File: public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee [hunk 5]
```diff
@@ -1,12 +1,83 @@
 define () ->
+	noArgumentCommands = [
```
Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: Wrong indentation

Comment (natestemen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: What exactly is wrong here?

Comment (natestemen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: Do you want each command on a new line?

Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: Hard to tell from Github, but I think it's indented with a tab, while the rest of the file is spaces?

Comment (jpallen) [inline public/coffee/ide/editor/directives/aceEditor/auto-complete/SnippetManager.coffee]: It's just the whitespace at the beginning that seems wrong, the list is fine 👍

**Comment by jpallen** [comment]
Just tested this locally, and its showing duplicates for commands that are already used in the document and also part of the static list:

screen shot 2017-08-15 at 09 43 38

It looks like this will be hard to fix with the code structured as-is, because the static list is in `SnippetsManager` and the dynamic list is in `SuggestionsManager`. It looks like `SnippetsManager` currently does some logic to pull out any environments used in the document already, while `SuggestionsManager` handles this for commands. I'd suggest then renaming:

* `SnippetsManager` -> `EnvironmentsManager`
* `SuggestionsManager` -> `CommandsManager`

and moving the static list of commands (but not environments) into the `CommandsManager`, where the logic for finding commands that are used in the document can easily reference the static list and ignore commands already included on that.

How does that sound? Sorry, these things have a habit of having more hidden complexity than expected for just adding a static list of commands!

**Comment by natestemen** [comment]
This sounds good. I will work on this today!

**Comment by jpallen** [comment]
New changes look good 👍 

Still got a problem with duplicate environments though:

screen shot 2017-08-17 at 14 02 48

One of these is a hardcoded snippet that includes and `\item` that gets added. The other is just a naive environment picked out from the document. I guess we want to blacklist the snippet environments from being included in the dynamic environments list?

Also, at the moment this conflicts with master which we'll need to sort before we can merge.

**Comment by natestemen** [comment]
Okay, I can work on this, but it is also a problem on the live website (not something introduced with this work).

Do you think we should take the snippet off the dynamic autocomplete when it appears, or if just don't add completions to the dynamic list if they are already in the snippet list. I feel like the snippets probably help enforce good practices.

**Comment by jpallen** [comment]
Yea, just don't add completions to the dynamic list if the snippet already exists 👍

**Comment by natestemen** [comment]
Another thing to do eventually is if we find `renewenvironment`, and in particular it redefines one of the environments we have a snippet for, we should remove it from the snippets.

**Comment by jpallen** [comment]
LGTM, just needs the conflicts with master resolving

**Pull request merged**